mssmt/tree.go

```diff
@@ -333,6 +333,112 @@ func (t *FullTree) MerkleProof(ctx context.Context, key [hashSize]byte) (
 	return NewProof(proof), nil
 }
 
+// findLeaves recursively traverses the tree represented by the given node and
+// collects all non-empty leaf nodes along with their reconstructed keys.
+func findLeaves(ctx context.Context, tx TreeStoreViewTx, node Node,
+	keyPrefix [hashSize]byte, depth int) (map[[hashSize]byte]*LeafNode, error) {
+
+	// Base case: If it's a leaf node.
+	if leafNode, ok := node.(*LeafNode); ok {
+		if leafNode.IsEmpty() {
+			return make(map[[hashSize]byte]*LeafNode), nil
+		}
+		return map[[hashSize]byte]*LeafNode{keyPrefix: leafNode}, nil
+	}
+
+	// Recursive step: If it's a branch node.
+	if branchNode, ok := node.(*BranchNode); ok {
+		// Optimization: if the branch is empty, return early.
+		if IsEqualNode(branchNode, EmptyTree[depth]) {
+			return make(map[[hashSize]byte]*LeafNode), nil
+		}
+
+		left, right, err := tx.GetChildren(depth, branchNode.NodeHash())
+		if err != nil {
+			return nil, fmt.Errorf("error getting children for "+
+				"branch %s at depth %d: %w",
+				branchNode.NodeHash(), depth, err)
+		}
+
+		// Recursively find leaves in the left subtree. The key prefix
+		// remains the same as the 0 bit is implicitly handled by the
+		// initial keyPrefix state.
+		leftLeaves, err := findLeaves(
+			ctx, tx, left, keyPrefix, depth+1,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		// Recursively find leaves in the right subtree. We need to set
+		// the bit corresponding to the current depth in the key prefix.
+		rightKeyPrefix := keyPrefix
+		byteIndex := depth / 8
+		bitIndex := depth % 8
+		rightKeyPrefix[byteIndex] |= (1 << bitIndex)
```

Comment:
I think that this should be spun into its own function so that we can add a bit more docs to explain what's happening here. I'm surprised that this function doesn't exist already. It's repeated in `PackBits` for example.